plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/RunTimeDependencyProcessor.xtend

```diff
@@ -0,0 +1,203 @@
+
+	/**
+	 * Invoked during AST traversal (and thus during main post-processing).
+	 */
+	def package recordRunTimeReferencesInCache(EObject node, ASTMetaInfoCache cache) {
+		if (node instanceof IdentifierRef) {
+			val target = node.targetElement;
+			if (N4JSLanguageUtils.hasRunTimeRepresentation(target, variantHelper)) {
+				cache.elementsReferencedAtRunTime += target;
+				// in case of namespace imports, we also want to remember that the namespace was referenced at run time:
+				val targetRaw = node.id;
+				if (targetRaw !== target && targetRaw instanceof ModuleNamespaceVirtualType) {
+					cache.elementsReferencedAtRunTime += targetRaw;
+				}
+			}
+		} else if (node instanceof N4ClassifierDeclaration) {
+			if (N4JSLanguageUtils.hasRunTimeRepresentation(node, variantHelper)) {
+				val targetTypeRefs = node.superClassifierRefs;
+				for (targetTypeRef : targetTypeRefs) {
+					val targetDeclType = targetTypeRef?.declaredType;
+					cache.elementsReferencedAtRunTime += targetDeclType;
+					val targetModule = if (targetDeclType !== null && !targetDeclType.eIsProxy) EcoreUtil2.getContainerOfType(targetDeclType, TModule);
+					if (isDifferentModuleInSameProject(targetModule, cache)) {
+						if (N4JSASTUtils.isTopLevelCode(node)) { // nested classifiers not supported yet, but let's be future proof
+							cache.modulesReferencedAtLoadTimeForInheritance += targetModule;
+						}
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Invoked at the end of AST traversal (and thus, during main post-processing).
+	 */
+	def package void storeDirectRunTimeDependenciesInTModule(Script script, ASTMetaInfoCache cache) {
+		val module = script.module;
+		val importDecls = script.scriptElements.filter(ImportDeclaration).toList;
+
+		// step 1: set run-time retention flag in import specifiers
+		for (importDecl : importDecls) {
+			for (importSpec : importDecl.importSpecifiers) {
+				if (importSpec !== null) {
+					val element = switch(importSpec) {
+						NamedImportSpecifier: importSpec.importedElement
+						NamespaceImportSpecifier: importSpec.definedType
+						default: throw new IllegalStateException("unknown subclass of ImportSpecifier: " + importSpec.eClass.name)
+					};
+					val retained = cache.elementsReferencedAtRunTime.contains(element);
+					EcoreUtilN4.doWithDeliver(false, [
+						importSpec.retainedAtRunTime = retained;
+					], importSpec);
+				}
+			}
+		}
+
+		// step 2: derive direct run-time dependencies from run-time retention of imports
+		//
+		// NOTE: order of the dependencies is important, here, because a change in the TModule has
+		// significant implications (e.g. dependent modules will be rebuilt, see #isAffected() in
+		// incremental builder); so we want to avoid a random reordering of the same dependencies and
+		// therefore use a defined order derived from the order of import declarations in the AST.
+		//
+		val modulesReferencedAtRunTime = newLinkedHashSet;
+		for (importDecl : importDecls) {
```

Comment:
It could be merged, but the performance gain would be marginal and then the two steps could not be separated as clearly for readability. Note that `importDecls` is a list, so it won't go twice through the script elements and perform the check if it's an import declaration.